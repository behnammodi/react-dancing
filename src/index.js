/* eslint-disable react/prop-types, react/display-name */
import { useRef, useLayoutEffect } from 'react';

const useDancer = ({
  delay = 0,
  duration = 200,
  defaultValue = 0,
  interpolate = {},
  timingFunction = (x) => x,
} = {}) => {
  const KEY_DANCER = 0;
  const KEY_RAF = 1;
  const KEY_VALUE = 2;
  const KEY_TIMEOUT = 3;
  const KEY_CURRENT_DANCER = 4;
  const { current: refs } = useRef(new Map());

  const getRef = (key) => refs.get(key);
  const setRef = (key, value) => refs.set(key, value);

  const setStyle = (style, element) => {
    for (let key in style) element.style[key] = style[key];
  };

  const raf = (handler) => setRef(KEY_RAF, requestAnimationFrame(handler));

  const setStyleByTranslatedValue = (value) => {
    const nextStyle = {};
    for (let key in interpolate)
      nextStyle[key] = interpolate[key](timingFunction(+value.toFixed(10)));

    setStyle(nextStyle, getRef(KEY_DANCER));
  };

  useLayoutEffect(() => {
    if (!getRef(KEY_DANCER)) return;

    if (getRef(KEY_DANCER) === getRef(KEY_CURRENT_DANCER)) return;

    setRef(KEY_CURRENT_DANCER, getRef(KEY_DANCER));

    setRef(KEY_VALUE, defaultValue);

    setStyleByTranslatedValue(defaultValue);
  });

  const start = (toValue) => {
    if (!getRef(KEY_DANCER)) return;

    if (toValue < 0 || toValue > 1) return;

    clearTimeout(getRef(KEY_TIMEOUT));
    setRef(
      KEY_TIMEOUT,
      setTimeout(() => {
        cancelAnimationFrame(getRef(KEY_RAF));
        const previousValue = getRef(KEY_VALUE);
        const isForward = toValue > previousValue;

        const previousTimestamp = previousValue === 0 ? 0 : duration * previousValue;

        let start;
        let currentValue = previousValue;
        function animate(timestamp) {
          if (start === undefined) {
            start = timestamp;
          } else {
            setStyleByTranslatedValue(currentValue);
          }

          const elapsed = timestamp - start;

          if (isForward) {
            if (currentValue < toValue) {
              currentValue = (elapsed + previousTimestamp) / duration;
              if (currentValue > toValue) currentValue = toValue;
              raf(animate);
              setRef(KEY_VALUE, currentValue);
            }
          } else {
            if (currentValue > toValue) {
              currentValue = (previousTimestamp - elapsed) / duration;
              if (currentValue < toValue) currentValue = toValue;
              raf(animate);
              setRef(KEY_VALUE, currentValue);
            }
          }
        }

        raf(animate);
      }, delay)
    );
  };

  const stop = () => {
    clearTimeout(getRef(KEY_TIMEOUT));
    cancelAnimationFrame(getRef(KEY_RAF));
  };

  return [(ref) => setRef(KEY_DANCER, ref), start, stop];
};

export { useDancer };
