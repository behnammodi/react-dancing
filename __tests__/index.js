/* eslint-disable no-undef */
import '@babel/polyfill';
import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import { Dancer, useDancer } from '../src';
import { create } from 'react-test-renderer';

const delay = (timeout) => new Promise((done) => setTimeout(done, timeout));

test('Snapshot a simple <Dancer />', () => {
  function App() {
    const [ref] = useDancer();
    return <Dancer ref={ref}></Dancer>;
  }

  const component = create(<App />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Snapshot a <Dancer /> with some props', () => {
  function App() {
    const [ref] = useDancer();
    return (
      <Dancer ref={ref} className="a" id="b" style={{ margin: 100 }}></Dancer>
    );
  }

  const component = create(<App />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Check style a <Dancer />', async () => {
  function App() {
    const [ref, play] = useDancer();

    useEffect(() => {
      play({ transform: 'translateX(100px)' });
    }, []);

    return (
      <Dancer ref={ref} className="a" id="b" style={{ margin: 100 }}></Dancer>
    );
  }
  const root = document.createElement('div');

  ReactDOM.render(<App />, root);

  const dancer = root.querySelector('div');

  /**
   * defualt style
   */
  expect(dancer.style.transitionDuration).toBe('0.2s');

  await delay(10);

  /**
   * style after run play
   */
  expect(dancer.style.transform).toBe('translateX(100px)');
});
