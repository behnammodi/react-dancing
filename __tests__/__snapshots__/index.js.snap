// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot a <Dancer /> with some props 1`] = `
<div
  className="a"
  id="b"
  style={
    Object {
      "margin": 100,
    }
  }
/>
`;

exports[`Snapshot a simple <Dancer /> 1`] = `<div />`;
